engine/src/main/java/io/camunda/zeebe/engine/processing/common/ElementActivationBehavior.java

```diff
@@ -0,0 +1,245 @@
+    this.keyGenerator = keyGenerator;
+    this.catchEventBehavior = catchEventBehavior;
+    this.elementInstanceState = elementInstanceState;
+    commandWriter = writers.command();
+    stateWriter = writers.state();
+  }
+
+  /**
+   * Activates the flow scopes of a given element. This is used when modifying a process instance or
+   * starting a process instance at a different place than the start event.
+   *
+   * @param processInstanceRecord the record of the process instance
+   * @param elementToActivate The element to activate
+   * @return The key of the activated element instance
+   */
+  public long activateElement(
+      final ProcessInstanceRecord processInstanceRecord,
+      final AbstractFlowElement elementToActivate) {
+
+    final var flowScopes = collectFlowScopesOfElement(elementToActivate);
+
+    final var flowScopeKey =
+        activateFlowScopes(
+            processInstanceRecord, processInstanceRecord.getProcessInstanceKey(), flowScopes);
+
+    final long elementInstanceKey =
+        activateElementByCommand(processInstanceRecord, elementToActivate, flowScopeKey);
+
+    // applying the side effects is part of creating the event subscriptions
+    sideEffectQueue.flush();
+
+    return elementInstanceKey;
+  }
+
+  private List<ExecutableFlowElement> collectFlowScopesOfElement(
+      final ExecutableFlowElement element) {
+    final List<ExecutableFlowElement> flowScopes = new ArrayList<>();
+
+    ExecutableFlowElement currentElement = element.getFlowScope();
+
+    while (currentElement != null) {
+      flowScopes.add(currentElement);
+
+      currentElement = currentElement.getFlowScope();
+    }
+
+    Collections.reverse(flowScopes);
+    return flowScopes;
+  }
+
+  private long activateFlowScopes(
+      final ProcessInstanceRecord processInstanceRecord,
+      final long flowScopeKey,
+      final List<ExecutableFlowElement> flowScopes) {
+
+    if (flowScopes.isEmpty()) {
+      return flowScopeKey;
+    }
+    final var flowScope = flowScopes.remove(0);
+
+    final List<ElementInstance> elementInstancesOfScope =
+        findElementInstances(flowScope, flowScopeKey);
+
+    if (elementInstancesOfScope.isEmpty()) {
+
+      final long elementInstanceKey =
+          activateFlowScope(processInstanceRecord, flowScopeKey, flowScope);
+      return activateFlowScopes(processInstanceRecord, elementInstanceKey, flowScopes);
+
+    } else if (elementInstancesOfScope.size() == 1) {
+      final var elementInstance = elementInstancesOfScope.get(0);
+      return activateFlowScopes(processInstanceRecord, elementInstance.getKey(), flowScopes);
+
+    } else {
+      // todo: deal with multiple flow scopes found without ancestor selection (#10008)
+      final var flowScopeId = BufferUtil.bufferAsString(flowScope.getId());
+      throw new UnsupportedOperationException(
+          "Found more than one flow scope instance with id '%s'.".formatted(flowScopeId));
+    }
```

Comment:
🔧 This entire section took me a good while to figure out what was going on. I think it'd help a lot to comment the different situations that are possible. E.g.

1. If empty it means we don't have an element instance for this flow scope yet, thus we activate it and continue with the remainder of the flow scopes.
2. If size == 1 it means there already exists 1 element instance so we don't have to activate anything and continue with the remainder of the flow scopes.
3. If size > 1 it means we found multiple elements instances and we don't know which one we should use. (this one is clear because of the comment and exception 😄)